sys/include/net/netopt.h

```diff
@@ -878,6 +905,133 @@ typedef enum {
+    netopt_rssi_t strength;         /**< Received signal strength */
+} netopt_scan_result_t;
+
+/**
+ * @brief   Static initializer for a @ref netopt_scan_result_t
+ *
+ * @param   ch      Scanned channel
+ * @param   str     Received signal strength
+ */
+#define NETOPT_SCAN_RESULT_INITIALIZER(ch, str)         \
+    (netopt_scan_result_t) {                            \
+        .channel = ch,                                  \
+        .strength = str,                                \
+    }
+
+/**
+ * @brief   Basic callback type on network scan @ref NETOPT_CONNECT
+ */
+typedef void (*netopt_on_scan_result_t) (const struct netopt_scan_result *res);
+
+/**
+ * @brief   Basic network scan request
+ */
+typedef struct netopt_scan_request {
+    netopt_channel_t channel;           /**< Channel to scan */
+    netopt_on_scan_result_t scan_cb;    /**< Scan result callback */
+} netopt_scan_request_t;
+
+/**
+ * @brief   Static initializer for a @ref netopt_scan_request_t
+ *
+ * @param   ch      Channel to be scanned
+ * @param   cb      Scan result callback
+ */
+#define NETOPT_SCAN_REQUEST_INITIALIZER(ch, cb)         \
+    (netopt_scan_request_t) {                           \
+        .channel = ch,                                  \
+        .scan_cb = (netopt_on_scan_result_t)cb,         \
+    }
+
+/**
+ * @brief   Basic network connect result
+ */
+typedef struct netopt_connect_result {
+    netopt_channel_t channel;           /**< Connected channel */
+} netopt_connect_result_t;
+
+/**
+ * @brief   Static initializer for a @ref netopt_connect_result_t
+ *
+ * @param   ch      Connected channel
+ */
+#define NETOPT_CONNECT_RESULT_INITIALIZER(ch)               \
+    (netopt_connect_result_t) {                             \
+        .channel = ch,                                      \
+    }
+
+/**
+ * @brief   Basic disconnect result
+ */
+typedef struct netopt_disconnect_result {
+    netopt_channel_t channel;           /**< Channel of the disconnected AP */
+} netopt_disconnect_result_t;
+
+/**
+ * @brief   Static initializer for a @ref netopt_disconnect_result_t
+ *
+ * @param   ch      Channel of the disconnected AP
+ */
+#define NETOPT_DISCONNECT_RESULT_INITIALIZER(ch)        \
+    (netopt_disconnect_result_t) {                      \
+        .channel = ch,                                  \
+    }
+
+/**
+ * @brief   Basic callback type on network connection @ref NETOPT_CONNECT
+ */
+typedef void (*netopt_on_connect_result_t) (const struct netopt_connect_result *res);
+
+/**
+ * @brief   Basic callback type on network disconnection @ref NETOPT_CONNECT
+ */
+typedef void (*netopt_on_disconnect_result_t) (const struct netopt_disconnect_result *res);
+
+/**
+ * @brief   Basic network connect request
+ */
+typedef struct netopt_connect_request {
+    netopt_on_connect_result_t conn_cb;         /**< On connect callback */
+    netopt_on_disconnect_result_t disconn_cb;   /**< On disconnect callback */
```

Comment:
I see some redundancy with `NETDEV_EVENT_LINK_UP`/`NETDEV_EVENT_LINK_DOWN` with those - could #17902 be a generic alternative here?

Reply:
An alternative maybe, but not sure.
An improvement for interested threads who did not request the action definately!
Giving a callback is not enforced, and also I would not exclude the case to send a message in the callback.
If #17902 gets merged before, I would also try with this, but I´m not sure if both ways are mutually exclusive.